lib/ipam/ipam.go

```diff
@@ -548,7 +549,44 @@ func (s *blockAssignState) findOrClaimBlock(ctx context.Context, minFreeIps int)
 	return nil, false, errors.New("failed to find or claim a block")
 }
 
-func (c ipamClient) autoAssign(ctx context.Context, num int, handleID *string, attrs map[string]string, requestedPools []net.IPNet, version int, host string, maxNumBlocks int, rsvdAttr *HostReservedAttr) ([]net.IPNet, error) {
+type IPAMAssignmentInfo struct {
+	NumRequested       int
+	NumAssigned        int
+	IPVersion          int
+	NumBlocksOwned     int
+	MaxNumBlocks       int
+	ExhaustedPools     []string
+	StrictAffinity     bool
+	NoFreeAffineBlocks bool
+	HostReservedAttr   *HostReservedAttr
+}
+
+func (i *IPAMAssignmentInfo) String() string {
+	if i == nil {
+		return ""
+	}
+
+	ret := fmt.Sprintf("Assigned %v out of %v requested IPv%v addresses", i.NumAssigned, i.NumRequested, i.IPVersion)
+
+	switch {
+	case i.NoFreeAffineBlocks && i.StrictAffinity:
+		ret += "; No more free affine blocks and strict affinity enabled"
```

Comment:
The nice golang way to do this is with `strings.Builder` https://golang.org/pkg/strings/#Builder

Reply:
Oh, TIL! Will use it, thanks!